cmd/utils.go

```diff
@@ -42,3 +49,212 @@ func getWorkerCount() int {
+	var outdatedPackages []gpsr.OutdatedPackage
+
+	for _, pkgDl := range availablePackages.Packages {
+
+		pkgName := pkgDl.Package.Package
+
+		if _, found := installed[pkgName]; found {
+			outdatedPackages = append(outdatedPackages, gpsr.OutdatedPackage {
+				Package:    pkgName,
+				OldVersion: installed[pkgName].Version,
+				NewVersion: pkgDl.Package.Version,
+			})
+		}
+	}
+	return outdatedPackages
+}
+
+func tagOldInstallations(fileSystem afero.Fs, libraryPath string, outdatedPackages []gpsr.OutdatedPackage) []UpdateAttempt {
+	var updateAttempts []UpdateAttempt
+
+	//Tag each outdated pkg directory in library with "__OLD__"
+	for _, pkg := range outdatedPackages {
+		updateAttempts = append(updateAttempts, tagOldInstallation(fileSystem, libraryPath, pkg))
+	}
+
+	return updateAttempts
+}
+
+func tagOldInstallation(fileSystem afero.Fs, libraryPath string, outdatedPackage gpsr.OutdatedPackage) UpdateAttempt {
+	packageDir := filepath.Join(libraryPath, outdatedPackage.Package)
+	taggedPackageDir := filepath.Join(libraryPath, "__OLD__" + outdatedPackage.Package)
+
+	err := RenameDirRecursive(fileSystem, packageDir, taggedPackageDir)
+
+	if err != nil {
+		log.WithField("package dir", packageDir).Warn("error when backing up outdated package")
+		log.Error(err)
+	}
+
+
+	return UpdateAttempt{
+		Package: outdatedPackage.Package,
+		ActivePackageDirectory: packageDir,
+		BackupPackageDirectory: taggedPackageDir,
+		OldVersion: outdatedPackage.OldVersion,
+		NewVersion: outdatedPackage.NewVersion,
+	}
+}
+
+
+func restoreUnupdatedPackages(fileSystem afero.Fs, packageBackupInfo []UpdateAttempt) {
+
+	if len(packageBackupInfo) == 0 {
+		return
+	}
+
+	//libraryDirectoryFsObject, _ := fs.Open(libraryPath)
+	//packageFolderObjects, _ := libraryDirectoryFsObject.Readdir(0)
+
+	for _, info := range packageBackupInfo {
+		_, err := fileSystem.Stat(info.ActivePackageDirectory) //Checking existence
+		if err == nil {
+
+			fileSystem.RemoveAll(info.BackupPackageDirectory)
+
+			log.WithFields(log.Fields{
+				"pkg": info.Package,
+				"old_version": info.OldVersion,
+				"new_version": info.NewVersion,
+			}).Info("successfully updated package")
+
+		} else {
+			log.WithFields(log.Fields{
+				"pkg": info.Package,
+				"old_version": info.OldVersion,
+				"new_version": info.NewVersion,
+			}).Warn("could not update package, restoring last-installed version")
+			err := RenameDirRecursive(fileSystem, info.BackupPackageDirectory, info.ActivePackageDirectory)
+			if err != nil {
+				log.WithField("pkg", info.Package).Error(err)
+			}
+		}
+	}
+}
+
+func RenameDirRecursive(fileSystem afero.Fs, oldPath string, newPath string) error {
+	err := CopyDir(fileSystem, oldPath, newPath )
+
+	if err != nil {
+		return err
+	}
+
+	err = fileSystem.RemoveAll(oldPath)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+//TODO: Move into goutils.
+func CopyDir(fs afero.Fs, src string, dst string) error {
```

Comment:
Agreed. Ultimately, I want to add this to your goutils package that we're using elsewhere. This was originally only for setting up tests, but because of the weirdness with fs.Rename(), I ended up promoting this to a real function. It needs cleaning, but I figure we'll do that (and add tests) when we transfer it between repos.